### ISSUE TITLE
Erasing GPT/MBR slowness

### ISSUE THREAD

**Issue by Sebastian-Roth** [FOGProject/fos#19]
As posted in the forums ([1](https://forums.fogproject.org/topic/11989/erasing-current-mbr-gpt-tables-taking-about-5-10-minutes), [2](https://forums.fogproject.org/topic/12169/erasing-current-mbr-gpt-table-issue), [3](https://forums.fogproject.org/topic/12144/erasing-mbr-gpt-quite-long), [4](https://forums.fogproject.org/topic/12370/it-takes-longer-time-more-than-5-minutes-erasing-current-mbr-gpt-tables), [5](https://forums.fogproject.org/topic/12104/capture-deploy-tasks-hang-on-erasing-saving-partition-tables)) we seem to have an issue where erasing old partition tables is taking a very long time (4-5 minutes compared to a couple of seconds).

So far we know that kernel 4.15.2 is not having the issue but 4.16.6 has. We started testing and so far it looks like all 4.15.x are good.

Great find by @Quazz: https://www.clonezilla.org/downloads/stable/changelog.php
> Clonezilla live 2.5.6-22
> ...
> * Downgrade the Linux kernel to 4.16.16-2 due to an issue of Linux kernel 4.17 that accesses local device very slow.

Though the kernel versions don't seem to match it seems like others are seeing similar things as well.

**Comment by Sebastian-Roth** [comment]
Alright, in our tests it turns out that 4.16.4 introduced the issue. @Quazz tested all the versions and found that 4.15.3 to 4.16.3 were all working nice and fast. 4.16.4 and later kernels show the issue. Working on bisect the kernel commits now. More tests will follow.

**Comment by Quazz** [comment]
Doing this in order of testing

001 - Slow
196 - Slow
100 - Slow
150 - Slow
050 - Slow

First commit (196) is this one https://github.com/torvalds/linux/commit/e09070c51b280567695022237e57c428e548b355 I think.

**Comment by Quazz** [comment]
Since Sebastian told me that he didn't experience slowness on his Oracle VirtualBox VMs, I figured I'd play around with some settings to see if maybe I can narrow down where to look.

When I disable IO-APIC (and thus only use one core), erasing goes fast again!

So it looks like the problem only shows up when you have multiple cores

**Comment by Quazz** [comment]
Interesting stuff, with IO-APIC back enabled, erasing goes slow as expected. 

However, if you then (wait 3-5 seconds when it gets 'stuck') press a key such as PageUP (and others that produce weird output on the console when pressing them (interestingely enough shift+pageup seems to work too) several times in a row then the task manages to complete in a timely fashion. 

I have no idea what that is, maybe something to do with the console? (since only when changes happen to the console does the task manage to complete, potentially this helps explain why debug mode is fine, too). 

Just a thought (unsubstantiated), but perhaps it's not the sgdisk command getting stuck, perhaps it doesn't even getting invoked yet at all because the console output is 'still going'. Don't have anything to back that up, but it's worth looking into maybe.

**Comment by Sebastian-Roth** [comment]
I got something horribly wrong when building those 196 kernel binaries. Not sure what exactly but. Sorry for that!

Talking to @Quazz about it I had the idea that this could be related to the kernel random number generator. It uses different types of user input to generate most valuable randomness. Keyboard input is one of them. You can hit pretty much any key on the keyboard and after 10-15 keys it got enough an finishes straight away. This is also why we don't see the issue in debug mode. Random number generator is already filled enough when we type in the commands I suppose. 

Now testing step by step...
26696cdda301830a16511391a3b1515c9b3b17fb (nr. 196) = fast
ab5860f5ce700bc4becc4d6abf01cc380c7ffe85 (nr. 035) = fast
1d0d9058215e75533f01fbb3db93621f142e1a3d (nr. 025) = slow

**Comment by leshik** [comment]
Does any workaround exist for this issue when using new kernels?

**Comment by Quazz** [comment]
The hang is caused by a lack of available entropy for the rng which is used by the erasing step. 

Currently you should either use kernel 4.15.2 (or older) or if you have to use the new kernels, you can generate entropy manually by pressing keys and moving the mouse at the station in question. 

While we did find the kernel commit that causes this to occur, I think it is best to patch (or wait for an update) on the packages. Preliminary testing looks good anyway.

**Comment by leshik** [comment]
@Quazz does the patch exist already? I was unable to find any PRs related to crng in FOG.

**Comment by Sebastian-Roth** [comment]
Had to fix a couple of things when updating to the latest Buildroot environment that took me a little while. Latest inits and kernels are now uploaded. To use those run:
```
sudo -i
cd /var/www/fog/service/ipxe
mv bzImage bzImage.orig
wget https://fogproject.org/kernels/bzImage
mv bzImage32 bzImage32.orig
wget https://fogproject.org/kernels/bzImage32
mv init.xz init.xz.orig
wget https://fogproject.org/inits/init.xz
mv init_32.xz init_32.xz.orig
wget https://fogproject.org/inits/init_32.xz
```

Closing this issue as fixed now!

**Comment by Sebastian-Roth** [comment]
@Quazz Haha, seems like we are running into the same kind of issue with 2019.02.1 but this time it's not util-linux's libuuid call but openssh ssh-keygen hanging!!

Seems like this [commit in openssh](https://github.com/openssh/openssh-portable/commit/afa6e79b76fb52a0c09a29688b5c0d125eb08302) later made it into Buildroot and now we see the same hang as we earlier had on "Erasing GPT/MBR ..." (hanging on uuid generation).

**Comment by Sebastian-Roth** [comment]
After an extensive debugging session I was able to find a nice solution to this problem. Fixed in 61abe486.

I tracked it down to where `ssh-keygen` calles `RAND_status` and that call seems to block on some machines, maybe just virtual machines. Adding `haveged` is working great to fill the entropy pool on bootup so ssh key generation is not hanging and the user does not have to add entropy by using the keyboard either.
